lib/open_api_spex/controller.ex

```diff
@@ -0,0 +1,153 @@
+  manner as ExDocs, so first line will be used as a `summary` and whole
+  documentation will be used as `description` field.
+
+  ### `parameters`
+
+  Parameters of the endpoint are defined by `:parameters` tag which should be
+  map or keyword list that is formed as:
+
+  ```
+  [
+    param_name: definition
+  ]
+  ```
+
+  Where `definition` is `OpenApiSpex.Parameter.t()` structure or map or keyword
+  list that accepts the same arguments.
+
+  ### `responses`
+
+  Responses are controlled by `:responses` tag and currently supports only
+  `application/json` responses. Responses must be defined as a map or keyword list
+  in form of:
+
+  ```
+  %{
+    200 => {"Response name", schema},
+    :not_found => {"Response name", schema}
+  }
+  ```
+
+  Where atoms are the same as `Plug.Conn.Status.code/1` values.
+
+  ### `requestBody`
+
+  Controlled by `:body` parameter and is defined as a tuple in form
+  `{description, schema}`.
+
+  ### `tags`
+
+  Tags are controlled by `:tags` attribute. In contrast to other attributes, this
+  one will also inherit all tags defined as a module documentation attributes.
+
+  ## Example
+
+  ```
+  defmodule FooController do
+    use #{inspect(__MODULE__)}
+
+    @moduledoc tags: ["Foos"]
+
+    @doc """
+    Endpoint summary
+
+    More docs
+    """
+    @doc parameters: [
+      id: [in: :path, type: :string, required: true]
+    ]
+    @doc responses: [
+      ok: {"Foo document", FooSchema}
+    ]
+    def show(conn, %{id: id}) do
+      # …
+    end
+  end
+  ```
+  '''
+
+  alias OpenApiSpex.Operation
+
+  defmacro __using__(_opts) do
+    quote do
+      @doc false
+      @spec open_api_operation(atom()) :: OpenApiSpex.Operation.t()
+      def open_api_operation(name),
+        do: unquote(__MODULE__).__api_operation__(__MODULE__, name)
+
+      defoverridable open_api_operation: 1
+    end
+  end
+
+  @doc false
+  @spec __api_operation__(module(), atom()) :: Operation.t()
+  def __api_operation__(mod, name) do
+    {mod_meta, summary, docs, meta} = get_docs(mod, name)
+
+    %Operation{
+      description: docs,
+      operationId: "#{inspect(mod)}.#{name}",
+      parameters: build_parameters(meta),
+      requestBody: build_request_body(meta),
+      responses: build_responses(meta),
+      summary: summary,
+      tags: Map.get(mod_meta, :tags, []) ++ Map.get(meta, :tags, [])
+    }
+  end
+
+  defp get_docs(module, name) do
+    {:docs_v1, _anno, _lang, _format, _module_doc, mod_meta, mod_docs} =
+      Code.fetch_docs(module)
+
+    {_, _, _, docs, meta} =
+      Enum.find(mod_docs, fn
+        {{:function, ^name, _}, _, _, _, _} -> true
+        _ -> false
+      end)
+
+    docs = Map.get(docs, "en", "")
+
+    [summary | _] = String.split(docs, ~r/\n\s*\n/, parts: 2)
+
+    {mod_meta, summary, docs, meta}
+  end
+
+  defp build_parameters(%{parameters: params}) do
+    for {name, options} <- params do
+      {location, options} = Keyword.pop(options, :in, :query)
```

Comment:
It would be nice to see an `Operation.parameter/1` added that takes a keyword list, so we don't have to convert to positional arguments here.

Reply:
@moxley in theory I could use `Kernel.struct/2` or `Kernel.struct!/2`

Reply:
That's what `Operation.parameter/5` does already. I'm just saying we should refactor Operation so that we can reuse that code.

Reply:
yeah... Openapi generators have `--remove-operation-id-prefix` but it doesn't work for every generator =/